sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2421,6 +2423,60 @@ multiply_with_overflow_check(T x, T y) {
     return x * y;
 }
 
+namespace matrix_ext = ext::oneapi::experimental::matrix;
+
+// Matrix type string to matrix_type enum value conversion
+// Note: matrix type strings are defined in template specialization for
+// convertTypeToMatrixTypeString above
+std::optional<matrix_ext::matrix_type>
+convertMatrixTypeStringMatrixTypeEnumValue(std::string MatrixTypeString) {
+  assert(!MatrixTypeString.empty() &&
+         "MatrixTypeString type string can't be empty. Check if required "
+         "template specialization for convertTypeToMatrixTypeString exists.");
+  if ("matrix_type::bf16" == MatrixTypeString)
+    return matrix_ext::matrix_type::bf16;
+  else if ("matrix_type::fp16" == MatrixTypeString)
+    return matrix_ext::matrix_type::fp16;
+  else if ("matrix_type::tf32" == MatrixTypeString)
+    return matrix_ext::matrix_type::tf32;
+  else if ("matrix_type::fp32" == MatrixTypeString)
+    return matrix_ext::matrix_type::fp32;
+  else if ("matrix_type::fp64" == MatrixTypeString)
+    return matrix_ext::matrix_type::fp64;
+  else if ("matrix_type::sint8" == MatrixTypeString)
+    return matrix_ext::matrix_type::sint8;
+  else if ("matrix_type::sint16" == MatrixTypeString)
+    return matrix_ext::matrix_type::sint16;
+  else if ("matrix_type::sint32" == MatrixTypeString)
+    return matrix_ext::matrix_type::sint32;
+  else if ("matrix_type::sint64" == MatrixTypeString)
+    return matrix_ext::matrix_type::sint64;
+  else if ("matrix_type::uint8" == MatrixTypeString)
+    return matrix_ext::matrix_type::uint8;
+  else if ("matrix_type::uint16" == MatrixTypeString)
+    return matrix_ext::matrix_type::uint16;
+  else if ("matrix_type::uint32" == MatrixTypeString)
+    return matrix_ext::matrix_type::uint32;
+  else if ("matrix_type::uint64" == MatrixTypeString)
+    return matrix_ext::matrix_type::uint64;
+  return std::nullopt;
+}
+
+bool isMatrixSupportedByHW(std::string MatrixTypeStrUser, size_t RowsUser,
+                           size_t ColsUser,
+                           matrix_ext::matrix_type MatrixTypeRuntime,
+                           size_t MaxRowsRuntime, size_t MaxColsRuntime,
+                           size_t RowsRuntime, size_t ColsRuntime) {
+  std::optional<matrix_ext::matrix_type> MatrixTypeUserOpt =
+      convertMatrixTypeStringMatrixTypeEnumValue(MatrixTypeStrUser);
+  return (MatrixTypeUserOpt ? (MatrixTypeUserOpt.value() == MatrixTypeRuntime)
+                            : false) &&
+         ((RowsRuntime != 0) ? (RowsUser == RowsRuntime)
+                             : (RowsUser <= MaxRowsRuntime)) &&
+         ((ColsRuntime != 0) ? (ColsUser == ColsRuntime)
+                             : (ColsUser <= MaxColsRuntime));
```

Comment:
To me, that is a bit hard to read.

I think that there is at least one early exit which we can outline:
```c++
if (!MatrixTypeUserOpt)
  return false;
return /* that condition with rows and columns */;
```